alien-module/control.lua

```diff
@@ -181,7 +176,25 @@ function update_modules_on_surface(surface, chunk_pos)
 			surface.create_entity {name = 'item-on-ground' , position = module_pos , stack = {name = current_module_name, count = item_count}}
 		end
 	end
-	
+
+end
+
+local function copy_surfaces()
+	global.surfaces = {}
+	for k,v in pairs(game.surfaces) do
+		global.surfaces[k] = v
+	end
+	global.current_surface_iter_index = next(global.surfaces, nil)
+end
+
+local function update_surface_iter_index()
+	--Inputting a nil value to next() will get the first table index. 
+	--Occasionally next will return nil, and we want to skip nil indexes
+	if global.surfaces == nil then copy_surfaces() end
```

Comment:
What happens when a surface is added or removed? this only catches the case when the global.surfaces variable is undefined

Reply:
Basically, if you load a save with the previous version of the mod, on_init will not execute-- originally I was going to initialize this table there. I wanted to find a way to guarantee that this table would be created in a way that was compatible with saves using previous versions of the mod. on_init didn't seem like the best place to put it. 

In the event that a surface is added, it will be skipped on the current pass of this cache. But at the end of this run, global.surfaces is cleared and recreated from game.surfaces. Testing it against the mod portal uplaoded save with ~80 surfaces, it took approximately 15-30 seconds to finish a pass through all surfaces, and UPS was ~59. So I think it won't take long for that surface to get added in with the others. 

I hope that I have enough sanity checks for a surface being removed that nothing bad will happen. I tried testing out the save uploaded to the mod portal and deleting a surface while there. But factorissimo threw an error before alien-module could.

Reply:
I think that this line 193 can be safely removed, since line 236 should always execute before this function.